src/components/ble/DfuService.cpp

```diff
@@ -124,9 +124,11 @@ int DfuService::WritePacketHandler(uint16_t connectionHandle, os_mbuf* om) {
                    bootloaderSize,
                    applicationSize);
 
-      // wait until SystemTask has finished waking up all devices
-      while (systemTask.IsSleeping()) {
-        vTaskDelay(50); // 50ms
+      // Wait until SystemTask has disabled sleeping
+      // This isn't quite correct, as we don't actually know
```

Comment:
I'm a bit worried about this comment. 
IIRC this "wait" loop is needed to ensure that the peripherals needed for DFU (the SPI flash, mainly) are awake and ready. If this is not the case, the DFU could fail which would prevent users from upgrading their watch.

Reply:
It's equivalent to the current implementation in terms of correctness, actually a little better as it checks specifically whether sleeping is disabled rather than whether the device is awake. The comment exists because it is possible for something else to disable sleeping, pass this check in DfuService, something else to enable sleeping and sleep being entered before SystemTask actually processes the DisableSleeping message from DFUService. It's an extremely unlikely race condition, and there is pretty much nothing we can do about it unless we have some way to check with SystemTask that the DisableSleeping message from DFU has been received (rather than from any other task)

I'm open to some kind of "transactional wakelock" where you can wait for the system task to approve your wakelock request before continuing. But the implementation sounds tricky

Reply:
I understand that this is not "worse" than before, so that's fine for me ;-)
Yeah, this "disable sleep" feature is much more complex than I originally expected!